WooCommerce/Classes/POS/Card Present Payments/CardPresentPaymentPreflightAdaptor.swift

```diff
@@ -0,0 +1,52 @@
+import Foundation
+import Combine
+import enum Yosemite.CardReaderDiscoveryMethod
+
+protocol CardPresentPaymentPreflightControllerFacade {
+    func attemptConnection(discoveryMethod: CardReaderDiscoveryMethod) async throws -> CardReaderPreflightResult
+}
+
+final class CardPresentPaymentPreflightAdaptor: CardPresentPaymentPreflightControllerFacade {
+    private let preflightController: CardPresentPaymentPreflightController
+
+    init(preflightController: CardPresentPaymentPreflightController) {
+        self.preflightController = preflightController
+    }
+
+    @MainActor
+    func attemptConnection(discoveryMethod: CardReaderDiscoveryMethod) async throws -> CardReaderPreflightResult {
+        async let preflightResult: CardReaderPreflightResult = try preflightController.readerConnection
+            .compactMap { $0 }
+            .eraseToAnyPublisher()
+            .async()
+
+        // This isn't a great async method... it would be better if it returned its result,
+        // but it actually returns before connection is finished.
+        // To get around this, we use the subscription above.
+        await preflightController.start(discoveryMethod: discoveryMethod)
+
+        return try await preflightResult
+    }
+}
+
+// From https://medium.com/geekculture/from-combine-to-async-await-c08bf1d15b77
+fileprivate extension AnyPublisher {
+    func async() async throws -> Output {
```

Comment:
Do you think it's worth making this available for other use cases? Like an extension like AnyPublisher+Async.

Reply:
Yeah, possibly. I wanted to prove it out here first though... I'm still not sure whether we'll have some side effects.